firebase-database/src/androidTest/java/com/google/firebase/database/DataTest.java

```diff
@@ -2740,6 +2744,186 @@ public void onComplete(DatabaseError error, boolean committed, DataSnapshot curr
+
+    try {
+      // null + increment
+      latch.register();
+      ref.setValue(0, ServerValue.increment(1));
+      // numeric priorities are always doubles
+      expectedPriorities.add(1.0);
+
+      // long + double
+      latch.register();
+      ref.setValue(0, ServerValue.increment(1.5));
+      expectedPriorities.add(2.5);
+
+      latch.awaitAdvanceInterruptibly(0, IntegrationTestValues.getTimeout(), MILLISECONDS);
+
+      // Checking types first makes failures much more obvious
+      assertEquals(expectedPriorities, foundPriorities);
+    } finally {
+      ref.removeEventListener(listener);
+      ref.getDatabase().goOnline();
+    }
+  }
+
+  @Test
+  public void testServerIncrementOverflowAndTypeCoercion() throws DatabaseException, TimeoutException, InterruptedException {
+    DatabaseReference ref = IntegrationTestHelpers.getRandomNode();
+    List<Object> foundValues = new ArrayList<>();
+    // Note: all numeric values will be long or double, so they must be cast before being inserted.
+    List<Object> expectedValues = new ArrayList<>();
+
+    // Phaser is the closest built-in to a bidrectional latch. We could use a semaphore with a fixed
+    // number of permits, but the test would be fragile since the permit count isn't closely related
+    // to the test cases.
+    final Phaser latch = new Phaser(0);
+
+    // Going offline ensures that local events get queued up before server events
+    ref.getDatabase().goOffline();
+
+    ValueEventListener listener = ref.addValueEventListener(new ValueEventListener() {
+      @Override
+      public void onDataChange(DataSnapshot snapshot) {
+        foundValues.add(snapshot.getValue());
+        latch.arrive();
+      }
+
+      @Override
+      public void onCancelled(DatabaseError error) {}
+    });
+
+    try {
+      // long + double = double
+      latch.bulkRegister(2);
+      ref.setValue(1);
+      ref.setValue(ServerValue.increment(1.5));
+      expectedValues.add((long) 1);
+      expectedValues.add(2.5);
+
+      // double + long = double
+      latch.bulkRegister(2);
+      ref.setValue(1.5);
+      ref.setValue(ServerValue.increment(1));
+      expectedValues.add(1.5);
+      expectedValues.add(2.5);
+
+      // long overflow = double
+      latch.bulkRegister(2);
+      ref.setValue(Long.MAX_VALUE - 1);
+      ref.setValue(ServerValue.increment(2));
+      expectedValues.add(Long.MAX_VALUE - 1);
+      expectedValues.add((double)Long.MAX_VALUE + 1.0);
+
+
+      // long underflow = double
+      latch.bulkRegister(2);
+      ref.setValue(Long.MIN_VALUE + 1);
+      ref.setValue(ServerValue.increment(-2));
+      expectedValues.add(Long.MIN_VALUE + 1);
+      expectedValues.add((double)Long.MIN_VALUE - 1.0);
+
+      latch.awaitAdvanceInterruptibly(0, IntegrationTestValues.getTimeout(), MILLISECONDS);
+
+      // Checking types first makes failures much more obvious
+      List<Class> expectedTypes = expectedValues.stream().map(Object::getClass).collect(toList());
```

Comment:
I don't think you can use the streaming APIs, since we can only use Java 7 APIs.

Reply:
Even in unit tests?

Reply:
I asked in the Android Contrib Chatroom for clarification:

- We can theoretically use Java 8 APIs in unit tests.
- We should not use them in Integration Tests, as they run on device. We currently run our tests only on Pixel 2, which is why they pass with this API reference. Ideally, we should also run them on a device that only meets the minimum requirements (see https://github.com/firebase/firebase-android-sdk/blob/master/transport/transport-runtime/transport-runtime.gradle#L21-L34 for an example of how we can configure that).

`DataTest` is an integration test (it's under androidTest and runs on device). We should therefore rewrite this part. I also think our codebase will be a little saner if we just avoid using Java 7 APIs consistently

Reply:
Fixed